neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -390,23 +488,36 @@ func QmpUnplugMemory(virtualmachine *vmv1.VirtualMachine) error {
 	}
 	defer mon.Disconnect()
 
-	// remove latest pc-dimm device
-	cmd := []byte(fmt.Sprintf(`{"execute": "device_del", "arguments": {"id": "%s"}}`, memoryDevices[plugged-1].Data.Id))
-	_, err = mon.Run(cmd)
-	if err != nil {
-		return err
-	}
-	// wait a bit to allow guest kernel remove memory block
-	time.Sleep(time.Second)
+	// run from last to first
+	var i int
+	var merr error
+	for i = plugged - 1; i >= 0; i-- {
+		// remove pc-dimm device
+		cmd := []byte(fmt.Sprintf(`{"execute": "device_del", "arguments": {"id": "%s"}}`, memoryDevices[i].Data.Id))
+		_, err = mon.Run(cmd)
+		if err != nil {
+			merr = errors.Join(merr, err)
+			continue
+		}
+		// wait a bit to allow guest kernel remove memory block
+		time.Sleep(time.Second)
 
-	// remove corresponding (latest) memdev object
-	cmd = []byte(fmt.Sprintf(`{"execute": "object-del", "arguments": {"id": "%s"}}`, strings.ReplaceAll(memoryDevices[plugged-1].Data.Memdev, "/objects/", "")))
-	_, err = mon.Run(cmd)
-	if err != nil {
-		return err
+		// remove corresponding memdev object
+		cmd = []byte(fmt.Sprintf(`{"execute": "object-del", "arguments": {"id": "%s"}}`, strings.ReplaceAll(memoryDevices[i].Data.Memdev, "/objects/", "")))
+		_, err = mon.Run(cmd)
+		if err != nil {
+			merr = errors.Join(merr, err)
+			continue
```

Comment:
yes, `time.Sleep` was quick and dirty solution. Need re-implement many QMP queries with subscription for events from qemu (using Go channels, timeouts and so on, as event goes async not as answer on request) instead of of sleep pauses. But I prefer do that in separate PR to not overload this one.

Reply:
added to follow ups list